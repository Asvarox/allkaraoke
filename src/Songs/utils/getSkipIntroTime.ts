import { seconds, Song, SongPreview } from 'interfaces';
import getSongFirstNoteMs from 'Songs/utils/getSongFirstNoteMs';
import isDev from 'utils/isDev';
import isE2E from 'utils/isE2E';

export const SKIP_INTRO_MS = isDev() || isE2E() ? 1_000 : 15_000;

export default function getSkipIntroTime(song: Song | SongPreview): seconds {
    return (song.videoGap ?? 0) + Math.floor((getSongFirstNoteMs(song) - SKIP_INTRO_MS) / 1000);
}
