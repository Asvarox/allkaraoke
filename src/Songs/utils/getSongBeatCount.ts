import { Song } from 'interfaces';
import isNotesSection from 'Songs/utils/isNotesSection';
import { getLastNoteEnd } from 'Songs/utils/notesSelectors';

export default function getSongBeatCount(song: Song) {
  return Math.max(
    ...song.tracks.map((track) => {
      const lastNoteSection = track.sections.filter(isNotesSection).at(-1);
      return lastNoteSection ? getLastNoteEnd(lastNoteSection) : 0;
    }),
  );
}
