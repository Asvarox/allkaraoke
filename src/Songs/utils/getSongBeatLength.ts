import { Song, SongPreview } from 'interfaces';
import memoize from '../../utils/memoize';

export default memoize(function getSongBeatLength(song: Song | SongPreview): number {
  return (60 / song.bpm / (song.bar ?? 4)) * 1000;
});
