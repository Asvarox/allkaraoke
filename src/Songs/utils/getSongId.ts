import { Song, SongPreview } from 'interfaces';
import { removeAccents } from '../../utils/clearString';

export default function getSongId(song: Pick<Song | SongPreview, 'artist' | 'title'> & { id?: string }) {
    return song.id
        ? song.id
        : removeAccents(`${song?.artist}-${song?.title}`)
              .toLowerCase()
              /// remove non-alphanumeric characters and replace spaces with dashes
              .replace(/[^a-z0-9\s-]/g, '')
              .replace(/\s+/g, '-')
              /// remove trailing dashes  &  remove dashes from the start
              .replace(/-+$/, '')
              .replace(/^-+/, '')
              .replace(/-+/g, '-');
}
