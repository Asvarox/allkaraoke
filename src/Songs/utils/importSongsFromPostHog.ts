import SongsService from 'Songs/SongsService';
import { Song } from 'interfaces';
import { importSongsFromPostHogBase } from './importSongsFromPostHogBase';

const importSongsFromPostHog = async () => {
  const posthogKey = sessionStorage.getItem('posthog_key') || prompt('Enter PostHog PAT');

  if (!posthogKey) {
    return;
  }
  sessionStorage.setItem('posthog_key', posthogKey);

  const defaultFrom = new Date();
  defaultFrom.setDate(defaultFrom.getDate() - 10);

  let from = prompt('Enter from', localStorage.getItem('posthog_from') || defaultFrom.toISOString());
  if (from === null) {
    return;
  }
  from = new Date(from).toISOString();

  localStorage.setItem('posthog_from', new Date().toISOString());
  const makeRequest = async (url: string) => {
    const response = await fetch(url, {
      headers: {
        Authorization: `Bearer ${posthogKey}`,
      },
    });

    if (!response.ok) {
      throw new Error(`Request failed with status ${response.status}`);
    }

    return response.json();
  };

  await importSongsFromPostHogBase(
    makeRequest,
    await SongsService.getIndex(true),
    [],
    async (song: Song) => {
      return SongsService.store(song);
    },
    from,
  );
};

export default importSongsFromPostHog;
