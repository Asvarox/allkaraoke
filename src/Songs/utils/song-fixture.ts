import { Song } from 'interfaces';

export const mulitrack: Song = {
  id: 'multitrack',
  artist: 'E2ETest',
  artistOrigin: undefined,
  title: 'Multitrack',
  video: 'W9nZ6u15yis',
  language: 'Polish',
  year: '1994',
  gap: 1000,
  bpm: 350,
  bar: 4,
  sourceUrl: undefined,
  author: undefined,
  authorUrl: undefined,
  edition: undefined,
  genre: undefined,
  lastUpdate: undefined,
  videoGap: undefined,
  previewStart: undefined,
  previewEnd: undefined,
  volume: undefined,
  realBpm: 100,
  unsupportedProps: ['#SOMEPROP: some value', '#SOMEPROP2: some value2'],
  tracks: [
    {
      sections: [
        {
          type: 'notes',
          start: 0,
          notes: [
            {
              type: 'rapstar',
              start: 5,
              length: 10,
              pitch: 9,
              lyrics: 'Second ',
            },
            {
              type: 'rap',
              start: 20,
              length: 10,
              pitch: 10,
              lyrics: 'Third ',
            },
          ],
        },
        {
          type: 'notes',
          start: 40,
          notes: [
            {
              type: 'freestyle',
              start: 50,
              length: 10,
              pitch: 11,
              lyrics: 'Fourth ',
            },
            {
              type: 'star',
              start: 80,
              length: 5,
              pitch: 9,
              lyrics: 'Seventh ',
            },
            {
              type: 'normal',
              start: 86,
              length: 4,
              pitch: 10,
              lyrics: 'Seventh ',
            },
          ],
        },
        {
          start: 100,
          end: 120,
          type: 'pause',
        },
        {
          type: 'notes',
          start: 130,
          notes: [
            {
              type: 'normal',
              start: 140,
              length: 10,
              pitch: 10,
              lyrics: 'Eight ',
            },
          ],
        },
      ],
    },
    {
      name: 'test track 2',
      sections: [
        {
          type: 'notes',
          start: 0,
          notes: [
            {
              type: 'normal',
              start: 5,
              length: 10,
              pitch: 11,
              lyrics: 'First ',
            },
            {
              type: 'normal',
              start: 16,
              length: 9,
              pitch: 9,
              lyrics: 'Second ',
            },
            {
              type: 'normal',
              start: 30,
              length: 10,
              pitch: 10,
              lyrics: 'Third ',
            },
          ],
        },
        {
          type: 'notes',
          start: 50,
          notes: [
            {
              type: 'normal',
              start: 60,
              length: 10,
              pitch: 11,
              lyrics: 'Fourth ',
            },
            {
              type: 'normal',
              start: 75,
              length: 10,
              pitch: 12,
              lyrics: 'Fifth ',
            },
            {
              type: 'normal',
              start: 90,
              length: 10,
              pitch: 9,
              lyrics: 'Seventh ',
            },
          ],
        },
        {
          start: 110,
          end: 130,
          type: 'pause',
        },
        {
          type: 'notes',
          start: 140,
          notes: [
            {
              type: 'normal',
              start: 150,
              length: 10,
              pitch: 9,
              lyrics: 'Eight ',
            },
            {
              type: 'normal',
              start: 165,
              length: 10,
              pitch: 10,
              lyrics: 'Nine ',
            },
          ],
        },
      ],
    },
  ],
};
