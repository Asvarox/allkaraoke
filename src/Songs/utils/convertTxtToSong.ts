import { ExtractOptional, NotesSection, Section, Song } from 'interfaces';
import { ValuesType } from 'utility-types';
import notFalsy from '../../utils/notFalsy';
import { generatePlayerChangesForTrack } from './generatePlayerChanges';
import getSongId from './getSongId';
import mergeTracks from './mergeTracks';

export const typesMap = {
  R: 'rap',
  F: 'freestyle',
  '*': 'star',
  ':': 'normal',
  G: 'rapstar',
} as const;

export const txtTypesMap: Record<ValuesType<typeof typesMap>, keyof typeof typesMap> = {
  rap: 'R',
  freestyle: 'F',
  star: '*',
  normal: ':',
  rapstar: 'G',
};

export const songTXTKeys = [
  'TRACKNAMES',
  'YEAR',
  'EDITION',
  'GENRE',
  'LANGUAGE',
  'VIDEOGAP',
  'PREVIEWSTART',
  'CREATOR',
  'TITLE',
  'ARTIST',
  'BPM',
  'GAP',
  'ID',
  'LASTUPDATE',
  'VOLUME',
  'REALBPM',
  'PREVIEWEND',
  'CREATORURL',
  'SOURCEURL',
  'ARTISTORIGIN',
  'VIDEOID',
  'ALLKARAOKE_ID',
  'ALLKARAOKE_LASTUPDATE',
  'ALLKARAOKE_VOLUME',
  'ALLKARAOKE_REALBPM',
  'ALLKARAOKE_PREVIEWEND',
  'ALLKARAOKE_CREATORURL',
  'ALLKARAOKE_SOURCEURL',
  'ALLKARAOKE_ARTISTORIGIN',
  'ALLKARAOKE_VIDEOID',
] as const;

export type knownSongTxtKeys = ValuesType<typeof songTXTKeys> | 'VIDEO' | 'P1' | 'P2';

function getUnknownProps(txt: string) {
  return txt
    .split('\n')
    .filter(
      (line) => line.startsWith('#') && !(songTXTKeys as unknown as string[]).includes(line.substring(1).split(':')[0]),
    );
}

function getPropertyValueFromTxt(txt: string, keys: knownSongTxtKeys[], type?: 'string'): string | undefined;
function getPropertyValueFromTxt(txt: string, keys: knownSongTxtKeys[], type?: 'array'): string[] | undefined;
function getPropertyValueFromTxt(txt: string, keys: knownSongTxtKeys[], type?: 'number'): number | undefined;
function getPropertyValueFromTxt(
  txt: string,
  keys: knownSongTxtKeys[],
  type: 'string' | 'number' | 'array' = 'string',
) {
  const getValue = (key: knownSongTxtKeys) => {
    const regex = new RegExp(`#${key}\\:(.*)`);

    const value = txt.match(regex)?.[1];
    if (value === undefined) return undefined;
    if (type === 'number') {
      return Number(value.replace(',', '.')) ?? 0;
    } else if (type === 'array') {
      const vals = value.split(', ');

      return vals;
    }
    return value;
  };

  return keys.map(getValue).filter(notFalsy).at(0);
}

function getEmbeddedYoutubeVideoId(txt: string) {
  const value = getPropertyValueFromTxt(txt, ['VIDEO']) ?? '';
  const regex = new RegExp(`v=(.*?)(,|$)`);
  return value.match(regex)?.[1];
}

const LINE_BREAK_RELATIVE_REGEXP = /- -?\d+ -?\d+/;

export function getVideoId(url: string) {
  try {
    const linkUrl = new URL(url);
    return linkUrl.searchParams.get('v') || false;
  } catch (e: any) {
    return false;
  }
}
function safeJsonParse<T extends any, DV extends any = T>(str: any, defaultValue: DV): T | DV {
  try {
    return JSON.parse(str);
  } catch (e) {
    return defaultValue;
  }
}

function getTrackNames(txt: string) {
  const p1 = getPropertyValueFromTxt(txt, ['P1']);
  const p2 = getPropertyValueFromTxt(txt, ['P2']);
  if (p1 || p2) return [p1, p2];
  const trackNames: string[] = safeJsonParse<string[]>(getPropertyValueFromTxt(txt, ['TRACKNAMES']), []);

  return trackNames;
}

export default function convertTxtToSong(
  text: string,
  videoLink?: string,
  author?: string,
  authorUrl?: string,
  sourceUrl?: string,
): Song {
  const trackNames = getTrackNames(text);

  const additionalData = {
    year: getPropertyValueFromTxt(text, ['YEAR']),
    artistOrigin: getPropertyValueFromTxt(text, ['ALLKARAOKE_ARTISTORIGIN', 'ARTISTORIGIN']),
    lastUpdate: getPropertyValueFromTxt(text, ['ALLKARAOKE_LASTUPDATE', 'LASTUPDATE']),
    edition: getPropertyValueFromTxt(text, ['EDITION']),
    genre: getPropertyValueFromTxt(text, ['GENRE']),
    videoGap: getPropertyValueFromTxt(text, ['VIDEOGAP'], 'number'),
    realBpm: getPropertyValueFromTxt(text, ['ALLKARAOKE_REALBPM', 'REALBPM'], 'number'),
    previewStart: getPropertyValueFromTxt(text, ['PREVIEWSTART'], 'number'),
    previewEnd: getPropertyValueFromTxt(text, ['ALLKARAOKE_PREVIEWEND', 'PREVIEWEND'], 'number'),
    volume: getPropertyValueFromTxt(text, ['ALLKARAOKE_VOLUME', 'VOLUME'], 'number'),
    author: getPropertyValueFromTxt(text, ['CREATOR']) ?? author,
    authorUrl: getPropertyValueFromTxt(text, ['ALLKARAOKE_CREATORURL', 'CREATORURL']) ?? authorUrl,
    sourceUrl: getPropertyValueFromTxt(text, ['ALLKARAOKE_SOURCEURL', 'SOURCEURL']) ?? sourceUrl,
  } satisfies ExtractOptional<Song>;

  if (additionalData.videoGap) additionalData.videoGap = Math.floor(additionalData.videoGap);
  const title = getPropertyValueFromTxt(text, ['TITLE']) ?? '';
  const artist = getPropertyValueFromTxt(text, ['ARTIST']) ?? '';
  const video =
    getPropertyValueFromTxt(text, ['ALLKARAOKE_VIDEOID', 'VIDEOID']) ?? getEmbeddedYoutubeVideoId(text) ?? '';

  const song: Song = {
    id: getPropertyValueFromTxt(text, ['ALLKARAOKE_ID', 'ID']) ?? getSongId({ title, artist }),
    title,
    artist,
    video,
    language: getPropertyValueFromTxt(text, ['LANGUAGE'], 'array') ?? [],
    bpm: Number(getPropertyValueFromTxt(text, ['BPM'])?.replace(',', '.') ?? 0),
    bar: 4,
    gap: Number(getPropertyValueFromTxt(text, ['GAP'])?.replace(',', '.') ?? 0),
    ...additionalData,
    tracks: [],
    unsupportedProps: getUnknownProps(text),
    mergedTrack: {} as any,
  };

  // Either ID is entered or try to parse link
  if (videoLink?.length === 11) {
    song.video = videoLink;
  } else if (videoLink) {
    try {
      const linkUrl = new URL(videoLink);
      song.video = linkUrl.searchParams.get('v') || 'Invalid link';
    } catch (e: any) {
      song.video = `Invalid link: ${e.message}`;
    }
  }

  let sections: Section[] = [];
  let lastStart = 0;

  text.split('\n').forEach((line) => {
    if (line.startsWith('#')) return;
    const split = line.split(' ');

    // Sometimes the txts are wrongly formatted and include breaks/notes that just overlap
    // The assumption here is that tracks should start and end roughly similarly, so if a next
    // track starts just as the previous ends, it's considered a wrong formatting issue rather
    // than a new track
    if (Number(split[1]) < lastStart * 0.75 && sections.length) {
      // new track (song is a duet)
      if (trackNames[song.tracks.length])
        song.tracks.push({ sections, name: trackNames[song.tracks.length], changes: [] });
      else song.tracks.push({ sections, changes: [] });
      sections = [];
    }

    if (LINE_BREAK_RELATIVE_REGEXP.test(line)) {
      // "Pause" or relative line break
      const [, start, end] = split;
      sections.push({ start: Number(start), end: Number(end), type: 'pause' });
      sections.push({ start: Number(end), type: 'notes', notes: [] });

      lastStart = Number(end);
    } else if (line.startsWith('-')) {
      // new section
      sections.push({ start: Number(line.split('-')[1].trim()), notes: [], type: 'notes' });
    } else if (Object.keys(typesMap).includes(line[0])) {
      if (sections.length === 0) {
        sections = [{ start: 0, notes: [], type: 'notes' }];
      }
      // Note
      const lastSection = sections[sections.length - 1];

      const [type, start, length, pitch, ...lyrics] = split;

      (lastSection as NotesSection).notes.push({
        type: typesMap[type as keyof typeof typesMap] ?? 'normal',
        start: Number(start),
        length: Number(length),
        pitch: Number(pitch),
        lyrics: lyrics.join(' '),
      });
      lastStart = Number(split[1]);
    }
  });

  if (trackNames[song.tracks.length]) song.tracks.push({ sections, name: trackNames[song.tracks.length], changes: [] });
  else song.tracks.push({ sections, changes: [] });

  song.tracks = song.tracks
    .map((track) => ({
      ...track,
      // get rid of sections with no notes
      sections: track.sections.filter((section) => section.type === 'pause' || section.notes.length),
      changes: [],
    }))
    .map((track) => ({
      ...track,
      changes: generatePlayerChangesForTrack(track, song),
    }));

  song.mergedTrack = mergeTracks(song.tracks, song);

  return song;
}
