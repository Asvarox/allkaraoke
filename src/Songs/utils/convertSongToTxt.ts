import { Note, Song } from 'interfaces';
import { invert } from 'lodash-es';
import isNotesSection from 'Songs/utils/isNotesSection';
import { ValuesType } from 'utility-types';
import { SongTXTKeys, typesMap } from 'Songs/utils/convertTxtToSong';

export const txtTypesMap = invert(typesMap) as any as Record<ValuesType<typeof typesMap>, keyof typeof typesMap>;

function toTxtValue(key: SongTXTKeys, value: string | string[] | number | undefined) {
    if (value === undefined) return undefined;
    return `#${key}:${
        typeof value === 'number'
            ? value.toString(10).replace('.', ',')
            : Array.isArray(value)
            ? value.join(', ')
            : value
    }`;
}

const notesToText = (notes: Note[]) =>
    notes
        .map((note) => `${txtTypesMap[note.type]} ${note.start} ${note.length} ${note.pitch} ${note.lyrics}`)
        .join('\n');

export default function convertSongToTxt(song: Song) {
    const parsedValues: Record<keyof Omit<Song, 'tracks' | 'bar'>, string | undefined> = {
        artist: toTxtValue('ARTIST', song.artist),
        title: toTxtValue('TITLE', song.title),
        author: toTxtValue('CREATOR', song.author),
        authorUrl: toTxtValue('CREATORURL', song.authorUrl),
        bpm: toTxtValue('BPM', song.bpm),
        realBpm: toTxtValue('REALBPM', song.realBpm),
        year: toTxtValue('YEAR', song.year),
        sourceUrl: toTxtValue('SOURCEURL', song.sourceUrl),
        previewStart: toTxtValue('PREVIEWSTART', song.previewStart),
        previewEnd: toTxtValue('PREVIEWEND', song.previewEnd),
        video: toTxtValue('VIDEOID', song.video),
        videoGap: toTxtValue('VIDEOGAP', song.videoGap),
        volume: toTxtValue('VOLUME', song.volume),
        genre: toTxtValue('GENRE', song.genre),
        language: toTxtValue('LANGUAGE', song.language),
        lastUpdate: toTxtValue('LASTUPDATE', song.lastUpdate),
        edition: toTxtValue('EDITION', song.edition),
        gap: toTxtValue('GAP', song.gap),
    };
    const trackNames = song.tracks.map((track) => track.name ?? null);

    const lines: string[] = [];
    song.tracks
        .map((track) => track.sections)
        .forEach(([first, ...sections]) => {
            lines.push(isNotesSection(first) ? notesToText(first.notes) : `- ${first.start} ${first.end}`);

            sections.forEach((section) =>
                lines.push(
                    isNotesSection(section)
                        ? `- ${section.start}\n${notesToText(section.notes)}`
                        : `- ${section.start} ${section.end}`,
                ),
            );
        });

    return Object.values(parsedValues)
        .concat(trackNames.filter(Boolean).length ? toTxtValue('TRACKNAMES', JSON.stringify(trackNames)) : [])
        .concat(lines)
        .concat('E')
        .filter((value) => value !== undefined)
        .join('\n');
}
