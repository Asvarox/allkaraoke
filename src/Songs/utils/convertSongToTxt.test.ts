import convertSongToTxt from 'Songs/utils/convertSongToTxt';
import { mulitrack } from 'Songs/utils/song-fixture';

describe('convertSongToTxt', () => {
  it('should properly convert the song', () => {
    expect(convertSongToTxt(mulitrack)).toMatchSnapshot();
  });

  it('should use known metadata keys over unsupported props', () => {
    const song = { ...mulitrack, unsupportedProps: ['#TITLE:not-expected'] };

    const result = convertSongToTxt(song);

    expect(result).toContain(`#TITLE:${song.title}`);
    expect(result).not.toContain(`#TITLE:not-expected`);
  });
});
