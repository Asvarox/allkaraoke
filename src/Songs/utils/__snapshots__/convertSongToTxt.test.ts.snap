// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertSongToTxt > should properly convert the song 1`] = `
"#ID:multitrack
#ARTIST:E2ETest
#TITLE:Multitrack
#BPM:350
#REALBPM:100
#YEAR:1994
#VIDEOID:W9nZ6u15yis
#LANGUAGE:Polish
#GAP:1000
#TRACKNAMES:[null,"test track 2"]
#SOMEPROP: some value
#SOMEPROP2: some value2
: 5 10 9 Second 
* 20 10 10 Third 
- 40
: 50 10 11 Fourth 
: 80 10 9 Seventh 
- 100 120
- 130
: 140 10 10 Eight 
: 5 10 11 First 
: 16 9 9 Second 
: 30 10 10 Third 
- 50
: 60 10 11 Fourth 
: 75 10 12 Fifth 
: 90 10 9 Seventh 
- 110 130
- 140
: 150 10 9 Eight 
: 165 10 10 Nine 
E"
`;
